src/main/java/org/wildfly/security/ssl/SSLUtils.java

```diff
@@ -103,23 +105,35 @@ private SSLUtils() {}
                 for (Provider.Service service : services) {
                     if (serviceType.equals(service.getType())) {
                         String protocolName = service.getAlgorithm();
-                        if (!preferredProviderByAlgorithm.containsKey(protocolName)) {
-                            preferredProviderByAlgorithm.put(protocolName, provider);
+                        List<Provider> providerList = preferredProviderByAlgorithm.get(protocolName);
+                        if (providerList == null) {
+                            providerList = new ArrayList<>();
+                            preferredProviderByAlgorithm.put(protocolName, providerList);
                         }
+                        providerList.add(provider);
                     }
                 }
             }
         }
 
-        // now figure out the supported protocol set.
-
-        String[] supportedProtocols = protocolSelector.evaluate(preferredProviderByAlgorithm.keySet().toArray(NO_STRINGS));
-        for (String supportedProtocol : supportedProtocols) {
-            Provider provider = preferredProviderByAlgorithm.get(supportedProtocol);
-            if (provider != null) {
-                return createSimpleSslContextFactory(supportedProtocol, provider);
-            }
+        // now return a factory that will return the best match is can create.
+        final String[] supportedProtocols = protocolSelector.evaluate(preferredProviderByAlgorithm.keySet().toArray(NO_STRINGS));
+        if (supportedProtocols.length > 0) {
+            return () -> {
+                for (String protocol : supportedProtocols) {
+                    List<Provider> providerList = preferredProviderByAlgorithm.get(protocol);
```

Comment:
Again you can save a smidge of code by using ```getOrDefault(protocol, emptyList())``` and inlining intermediate variables accordingly.